File: Algorithms/src/3720. Lexicographically Smallest Permutation Greater Than Target/Code/solution.py
class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        # Store the frequency of every character available in s.
        count = [0] * 26
        for ch in s:
            count[ord(ch) - ord('a')] += 1

        n = len(s)
        matched = 0

        # Match target from left to right for as long as possible.
        while matched < n and count[ord(target[matched]) - ord('a')] > 0:
            # Using the same character keeps the current prefix smallest.
            count[ord(target[matched]) - ord('a')] -= 1
            matched += 1

        # Start where matching failed, or at the last position if all matched.
        start = matched if matched < n else n - 1

        # Move backward because increasing a later position gives a smaller answer.
        for i in range(start, -1, -1):
            # Restore the character if it was previously used to match target.
            if i < matched:
                count[ord(target[i]) - ord('a')] += 1

            # Find the smallest available character greater than target[i].
            bigger = -1
            for ch in range(ord(target[i]) - ord('a') + 1, 26):
                if count[ch] > 0:
                    bigger = ch
                    break

            # Build the answer as soon as this position can be increased.
            if bigger != -1:
                # Consume the character used to make the answer greater.
                count[bigger] -= 1

                # Keep the prefix unchanged and place the larger character.
                answer = target[:i] + chr(ord('a') + bigger)

                # Append all remaining characters in sorted order.
                for ch in range(26):
                    answer += chr(ord('a') + ch) * count[ch]

                return answer

        # No permutation is strictly greater than target.
        return ""

File: Algorithms/src/3720. Lexicographically Smallest Permutation Greater Than Target/Code/solution.go
func lexGreaterPermutation(s string, target string) string {
	// Store the frequency of every lowercase character available in s.
	count := make([]int, 26)
	for i := 0; i < len(s); i++ {
		count[s[i]-'a']++
	}

	n := len(s)
	matched := 0

	// Match target from left to right for as long as possible.
	for matched < n && count[target[matched]-'a'] > 0 {
		// Using the same character keeps the current prefix smallest.
		count[target[matched]-'a']--
		matched++
	}

	// Start where matching failed, or at the last position if all matched.
	start := n - 1
	if matched < n {
		start = matched
	}

	// Try to increase the answer from right to left.
	for i := start; i >= 0; i-- {
		// Restore this character if it was previously used to match target.
		if i < matched {
			count[target[i]-'a']++
		}

		// Find the smallest available character strictly greater than target[i].
		bigger := -1
		for ch := int(target[i]-'a') + 1; ch < 26; ch++ {
			if count[ch] > 0 {
				bigger = ch
				break
			}
		}

		// Build the answer when an increasing character is found.
		if bigger != -1 {
			// Consume the character that makes the string strictly greater.
			count[bigger]--

			// Keep the prefix before i unchanged.
			answer := make([]byte, 0, n)
			answer = append(answer, target[:i]...)

			// Add the smallest available character greater than target[i].
			answer = append(answer, byte('a'+bigger))

			// Append all remaining characters in sorted order.
			for ch := 0; ch < 26; ch++ {
				for count[ch] > 0 {
					answer = append(answer, byte('a'+ch))
					count[ch]--
				}
			}

			return string(answer)
		}
	}

	// No permutation is strictly greater than target.
	return ""
}